TestRunner.forceImmediateCompletion() hangs with site isolation

kkinnunen-apple opened the pull request:
#### 34863c0ae258b67d0817a9cbde6086426e0602e7
<pre>
TestRunner.forceImmediateCompletion() hangs with site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=280403">https://bugs.webkit.org/show_bug.cgi?id=280403</a>
<a href="https://rdar.apple.com/136748844">rdar://136748844</a>

Reviewed by Alex Christensen.

TestRunner.forceImmediateCompletion() would skip work because
it would try to run InjectedBundlePage::dump() in the calling WCP.
This would try to access main frame url, which is not available as
it is a remote frame.

Fix by resolving whether to dump or not by asking the WKTR process.

This is incorrect for forceImmediateCompletion(), as the semantics is
that the dump happens at that JS call stack immediately. This of
course is infeasible for site isolation, as the dump is global but
the processes cannot re-enter like that.

This is also incorrect for earlier notifyDone(). The notifyDone()
should dump immediately in case the load has happened.
(bug 268471 274211@main)

Add tests for WKTR to explain the semantics. These semantics should
likely change, as they&apos;re incompatible with how site isolation works.
The re-entrant semantics has been causing mismatches also
without site isolation, so it may be beneficial to rethink these.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-async-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-async.html: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-sync-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-sync.html: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-sync-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-sync.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-async-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-async.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-sync-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-sync.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-sync-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-sync.html: Added.
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::notifyDone):
(WTR::TestRunner::forceImmediateCompletion):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
(WTR::TestInvocation::resolveNotifyDone):
(WTR::TestInvocation::resolveForceImmediateCompletion):
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::siteIsolationEnabled const):

Canonical link: <a href="https://commits.webkit.org/285592@main">https://commits.webkit.org/285592@main</a>
</pre>

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
```diff
@@ -274,6 +274,14 @@ void TestRunner::forceImmediateCompletion()
         return;
 
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    bool mainFrameIsRemote = WKBundleFrameIsRemote(WKBundlePageGetMainFrame(injectedBundle.pageRef()));
```
We don't want to increase use of WKBundleFrameIsRemote.  We should instead be moving state to the UI process and sending a sync message, the reply to which indicates if we should dump now.  It probably needs to be tied with TestRunner::notifyDone.

achristensen07 reviewed (approved):
This is a step in the right direction.  I think we'll want to remove the if (m_options.siteIsolationEnabled()) logic eventually.

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
```diff
@@ -250,34 +250,23 @@ void TestRunner::notifyDone()
-    bool mainFrameIsRemote = WKBundleFrameIsRemote(WKBundlePageGetMainFrame(injectedBundle.pageRef()));
-    ALLOW_DEPRECATED_DECLARATIONS_END
-    if (mainFrameIsRemote) {
-        setWaitUntilDone(false);
```
Can we remove this state from the web process now?  Same with shouldWaitUntilDone

kkinnunen-apple committed:
TestRunner.forceImmediateCompletion() hangs with site isolation
https://bugs.webkit.org/show_bug.cgi?id=280403
rdar://136748844

Reviewed by Alex Christensen.

TestRunner.forceImmediateCompletion() would skip work because
it would try to run InjectedBundlePage::dump() in the calling WCP.
This would try to access main frame url, which is not available as
it is a remote frame.

Fix by resolving whether to dump or not by asking the WKTR process.

This is incorrect for forceImmediateCompletion(), as the semantics is
that the dump happens at that JS call stack immediately. This of
course is infeasible for site isolation, as the dump is global but
the processes cannot re-enter like that.

This is also incorrect for earlier notifyDone(). The notifyDone()
should dump immediately in case the load has happened.
(bug 268471 274211@main)

Add tests for WKTR to explain the semantics. These semantics should
likely change, as they're incompatible with how site isolation works.
The re-entrant semantics has been causing mismatches also
without site isolation, so it may be beneficial to rethink these.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-async-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-async.html: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-sync-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-onload-sync.html: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-sync-expected.txt: Added.
* LayoutTests/wktr/force-immediate-completion-moment-dump-as-text-sync.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-async-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-async.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-sync-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-onload-sync.html: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-sync-expected.txt: Added.
* LayoutTests/wktr/notify-done-moment-dump-as-text-sync.html: Added.
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::notifyDone):
(WTR::TestRunner::forceImmediateCompletion):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
(WTR::TestInvocation::resolveNotifyDone):
(WTR::TestInvocation::resolveForceImmediateCompletion):
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::siteIsolationEnabled const):

Canonical link: https://commits.webkit.org/285592@main
